ytypes/string_type.go

```diff
@@ -26,6 +27,57 @@ import (
 
 // Refer to: https://tools.ietf.org/html/rfc6020#section-9.4.
 
+var (
+	// posixRegexCache stores previously-compiled Regexp objects.
+	// This helps the performance of validation of, say, a large prefix
+	// list that have the same pattern specification.
+	//
+	// Concurrency Requirements
+	//
+	// Only the regexp cache has to be protected by mutexes, since
+	// a Regexp is safe for concurrent use by multiple goroutines:
+	// https://golang.org/src/regexp/regexp.go
+	posixRegexCache = map[string]*regexp.Regexp{}
```

Comment:
I'd make these a type:

```golang
type regexpCache struct {
   posixMu sync.RWMutex 
   posix map[string]*regexp.Regexp
   
   re2Mu sync.RWMutex
   re2 map[string]*regexp.Regexp
}
```

This seems cleaner to me in terms of it being clearer what the mutexes are designed to guard (similar to the comment made w.r.t mutexes in [Effective Go](https://golang.org/doc/effective_go#commentary)).

Reply:
Done.